software/contrib/master_clock.py

```diff
@@ -307,34 +391,38 @@ def loadState(self):
 
         self.bpm = self.state.get("bpm", 100)
         self.pulseWidthPercent = self.state.get("pulseWidthPercent", 50)
+        self.externalClockInput = self.state.get("externalClockInput", False)
         self.outputDivisions = self.state.get("outputDivisions", [1,2,4,8,16,32])
 
         self.saveState()
 
     async def main(self):
         while True:
-
-            # Display selected screen
-            if self.screen == 1:
-                self.screen1()
-            else:
-                self.screen2()
+            if not self.clockSelectionScreenActive:
+                # Display selected screen
+                if self.screen == 1:
+                    self.screen1()
+                else:
+                    self.screen2()
 
             # Auto reset function after resetTimeout
             if self.step != 0 and ticks_diff(ticks_ms(), self.previousStepTime) > self.resetTimeout:
                  self.step = 1
                  self.completedCycles = 0
 
-            if not self.configMode:
+            if not self.configMode and not self.externalClockInput:
                 self.checkForAinBPM()
 
-            if self.running:
+            if self.running and not self.externalClockInput:
                 self.clockTrigger()
                 self.calcSleepTime()
                 if self.configMode:
-                    await asyncio.sleep_ms(int(self.timeToSleepMs - self.msDriftCompensation - self.msDriftCompensationConfigMode))
+                    await asyncio.sleep_ms(int(self.mSBetweenClockCycles - self.msDriftCompensation - self.msDriftCompensationConfigMode))
                 else:
-                    await asyncio.sleep_ms(int(self.timeToSleepMs - self.msDriftCompensation))
+                    await asyncio.sleep_ms(int(self.mSBetweenClockCycles - self.msDriftCompensation))
+            else:
+                # need to add this otherwise the async tasks never start - strange eh!
+                await asyncio.sleep_ms(1)
```

Comment:
[discussion] I _think_ that you want `sleep_ms(0)` here as your goal is to simply allow the other tasks to run. You don't actually care how long you sleep for.

https://docs.python.org/3/library/asyncio-task.html#sleeping

As to the comment above, I believe that the reason you need this is that otherwise you are in a busy while loop. You need to yield execution to other tasks if you actually want them to run.

Reply:
sleep_ms(0) works great - thanks for the info on why it's needed. Updated code.